### ISSUE TITLE
Make threshold more memory efficient

### ISSUE THREAD

**Issue by bayliffe** [metoppv/improver#1787]
## The issue
The threshold plugin currently processes multi realization data without any kind of slicing to reduce the amount of data being held in memory. This can lead to enormous memory requirements when processing data from models that produce large numbers of ensemble members (e.g. ECMWF). This is necessary if the realizations are being preserved, but when we are collapsing the realization coordinate, which is common, we could reduce the memory usage considerably.

## The work
This work is to re-engineer the threshold plugin to be more memory efficient in these cases, ideally showing consistent memory usage regardless of the number of ensemble members being processed.

## Follow on
We may need to undertake similar work for any other improver processing steps that don't use realization slicing and are thus likely to show undesirable memory behaviour when using larger ensembles.

Acceptance criteria:
- the threshold plugin shows more consistent memory usage when being applied to data with differing numbers of realizations when the realization coordinate is being collapsed; this should be demonstrated with statistics.

**Comment by mspelman07** [comment]
I have put up a draft PR (PR #1842) which slices the cube within the threshold plugin if the realizations are to be collapsed. 
Currently this doesn't seem to be reducing the amount of maximum memory used by the plugin.

**MAX-tracker for my updated threshold plugin:**

![Image](https://user-images.githubusercontent.com/99179165/204332710-7656f43a-f913-4863-b612-0983bfda159b.png)

**MAX-tracker for the current threshold plugin:**


![Image](https://user-images.githubusercontent.com/99179165/204333304-d06d623b-90a5-4e5b-9e62-1e260dba0551.png)

**Comment by mspelman07** [comment]
I have tried slicing over the cube in the threshold plugin to reduce the memory (as in the PR above).

Following discussion with @bayliffe he has determined that in its current state this does not reduce the memory requirements and in fact increases them (mainly noticeable when using cubes with many realizations and many thresholds). Partially this is due to the data being realised at the beginning of the plugin although this can be rectified to include the checks within the loops. 

Issues then arise due to the order of the “for” loops. The current order of the loops is such that it loops over the threshold first and for each threshold slices over each realization. This leads to a reduction in memory but massively slows dowen the plugin due to having to read the same slice from disk multiple times for each threhsold.

**Comment by bayliffe** [comment]
## Progress

I have managed to reduce the memory usage of the threshold CLI by about 1/3.

## Code changes
https://github.com/metoppv/improver/compare/master...bayliffe:improver:im1787?expand=1

### General approach
The changes are designed to avoid the need to hold a thresholded, multi-realization array in memory. At present the threshold plugin will threshold each realization in turn, and if collapsing realizations, perform an iris.analysis.MEAN over that realization dimension. The rewritten version, when collapsing realizations, will sum up the contributions to each threshold from each realization, and then divide by the total number of contributions. More specifically, at each point, from each realization the truth value is calculated and accumulated. The denominator is incremented by 1 for each contribution; this does not increase if the point is masked in a particular realization. In cases where the realization is not being collapsed, the rewritten plugin is unlikely to offer any benefits.

### Vicinity processing
To preserve vicinity functionality, the OccurenceWithinVicinity plugin has been broken up somewhat, so that key components can be called from within the BasicThreshold plugin incrementally. Currently the threshold plugin performs vicinity calculations on a per threshold per realization basis after all the thresholding is complete, and then any realization collapsing required is applied afterwards. If we want to avoid the need for a large threshold x realization array, we have to perform the vicinity processing for each threshold when we calculating the truth value, prior to adding it to the sum. Hence the need to rewrite this quite extensively.

### CLI
As part of these changes I have moved the processing / checking that was being done in the CLI into the plugin. This better aligns with our aims of having very simple (and one day redundant) CLIs. I've removed the ability to provide generic functions to the plugin, and instead limited the options to realization collapsing, which is what we actually do.

### Tests
I have removed redundant tests, e.g. testing the provision of generic functions to the threshold plugin. I have ensured that all other unit and acceptance tests pass unchanged (except for modified argument names / approaches).

## Testing
I have performed a majority of my testing using an ECMWF screen temperature diagnostic with 51 realizations. I have thresholded this data with out standard set of 71 threshold values. To test the vicinity processing I have used a 3 realization ENUKX screen temperature file.

|Experiment | Plugin | Highest step memory (MB) | Peak memory (MB) | Total run time (s) |
|---|---|---|---|---|
|Threshold, realization collapse|Current | 655 | 1882 | 113 |
|Threshold, realization collapse|Rewritten | 415 | 1285 | 84 |
|Threshold, vicinity, realization collapse|Current| 362| 1203 | 51 |
|Threshold, vicinity, realization collapse|Rewritten | 364 | 1448| 19 |

The outputs of these test are identical according to nccmp.
The vicinity test is interesting. There are only 3 realizations in this input file, so the memory would expect to be relatively similar, and that is the case for the step memory. The rewritten code appears to have a higher peak memory usage, which is odd. The peak memory usage in the vicinity case is of the same order as the memory when processing the 51 member EC file, which is somewhat surprising; Iris / dask / numpy are all doing things in here, so I have no idea why this is the case.

What is perhaps more surprising is that the rewritten code is considerably faster than the original. This wasn't the aim, which was purely to reduce memory use. This is likely a result of working with numpy arrays and avoiding constructing and then slicing over cubes, as must be done for vicinity processing in the current approach.

## Questions
1. The memory usage seen above is not wildly different between processing the ENUKX and EC data, so why was the suite having such issues with the thresholding?
2. Are the improvements seen here sufficient to warrant rewriting this core IMPROVER functionality? Note that the branch at present is not in a working state as lots of other code pulls in / hangs off of the thresholding and this has not yet been fixed; there is little point if we don't want to adopt this.
3. Should we hand this over to some of our computer science colleagues for better optimisation of the algorithm than I can provide?

**Comment by bayliffe** [comment]
I've moved this into review so we can discuss it. It will either go back into "In progress" for further work, or a new approach will be devised.

**Comment by MoseleyS** [comment]
I've given this a bit of thought and I think the reason we don't see any memory reduction is because the method still has to load the whole input array (r, y, x) at once.

What we need is a method that loops over the realization coordinate. Something like
```
thresholded_data = np.zeros(len(threshold_list), len(y), len(x), dtype=np.float32)
for realization_slice in cube.slices("realization"):
    # This loads just one realization_slice, which is discarded when the outer loop iterates
    for threshold in threshold_list:
        thresholded_slice = ...threshold realization_slice into boolean
        contribution = thresholded_slice * 1 / number_of_realizations

        # So now we just add the contribution from this realization for this threshold to the appropriate slice in the thresholded_data :
        thresholded_data [threshold].data += contribution
```

**Comment by MoseleyS** [comment]
Ben has pointed out that this is exactly what he has tried. Looking more closely at the code, I've spotted that `input_slices` is a list, so each item in the list is preserved while the code iterates through them. Changing this to be the `SliceIterator` inside the list reduces the memory requirement by 30% in one test. This does also require a small adjustment to how the empty `thresholded_cube` is defined, but that is trivial.

I used ECMWF data - Ben's version used 617MB, mine 430MB.

**Comment by bayliffe** [comment]
## Updates

### Further investigation

Katie looked at the memory usage of the threshold step when working with ECMWF data within the suite here: https://github.com/metoppv/mo-blue-team/issues/499
She demonstrated that the processing and memory requirements when thresholding the ECMWF data were higher than when working with the smaller ENGL ensemble, as might be expected:
```
ENGL took 60.76 seconds to run with a max RSS of 1661.70 MiB
ECGL took 125.149 seconds to run with a max RSS of 1971.56 Mib
```

### Stephen's suggestions

I have rebased the branch linked far above onto current master and I have folded in Stephen's changes to use an iterator, rather than a list, which he showed to reduce the memory usage. I repeated some experiments to confirm this improvement once again.

Thresholding an ECMWF screen temperature cube, using our standard 71 thresholds, the IMPROVER memory profiling returned the following results:

|Experiment | Plugin | Realizations | Peak memory (MB) | Total run time (s) |
|---|---|---|---|---|
|Threshold, realization collapse| Using List| 51| 1280 | 72 |
|Threshold, realization collapse| Using Iterator | 51| 1034 | 53|


### Suite test

I ran an [ECGL suite](http://fcm1/cylc-review/taskjobs/bayliffe?&suite=imecgl_mem_test) processing PMSL. I ran this using the master version of IMPROVER which [failed as expected](http://fcm1/cylc-review/view/bayliffe?&suite=imecgl_mem_test&no_fuzzy_time=0&path=log/job/20230614T1200Z/impr_ecgl_pressrhlight_PT0024H/01/job.err) for exceeding the memory available.

I then reloaded this suite and reinstalled IMPROVER containing the rewritten version of the threshold plugin linked above. Once this was done the suite ran successfully, [completing the thresholding and subsequent tasks](http://fcm1/cylc-review/view/bayliffe?&suite=imecgl_mem_test&no_fuzzy_time=0&path=log/job/20230614T1200Z/impr_ecgl_pressrhlight_PT0024H/03/job.out).

This is encouraging, and suggests the improvements made in this branch should be sufficient to allow us to run IMPROVER using ECMWF data, though future technical hurdles are far from ruled out.

## Next steps

I've rebased this long out of date branch on current master, and dealt with the resulting conflicts. The tests directly associated with basic thresholding all pass.

**Comment by bayliffe** [comment]
## PR: https://github.com/metoppv/improver/pull/1913

**Comment by bayliffe** [comment]
## Updates

### Further investigation

Katie looked at the memory usage of the threshold step when working with ECMWF data within the suite here: https://github.com/metoppv/mo-blue-team/issues/499
She demonstrated that the processing and memory requirements when thresholding the ECMWF data were higher than when working with the smaller ENGL ensemble, as might be expected:
```
ENGL took 60.76 seconds to run with a max RSS of 1661.70 MiB
ECGL took 125.149 seconds to run with a max RSS of 1971.56 Mib
```

### Stephen's suggestions

I have rebased the branch linked far above onto current master and I have folded in Stephen's changes to use an iterator, rather than a list, which he showed to reduce the memory usage. I repeated some experiments to confirm this improvement once again.

Thresholding an ECMWF screen temperature cube, using our standard 71 thresholds, the IMPROVER memory profiling returned the following results:

|Experiment | Plugin | Realizations | Peak memory (MB) | Total run time (s) |
|---|---|---|---|---|
|Threshold, realization collapse| Using List| 51| 1280 | 72 |
|Threshold, realization collapse| Using Iterator | 51| 1034 | 53|


### Suite test

I ran an ECGL suite processing PMSL. I ran this using the master version of IMPROVER which failed as expected for exceeding the memory available.

(Suite run under my user name, called imecgl_mem_test)

I then reloaded this suite and reinstalled IMPROVER containing the rewritten version of the threshold plugin linked above. Once this was done the suite ran successfully, completing the thresholding and subsequent tasks.

This is encouraging, and suggests the improvements made in this branch should be sufficient to allow us to run IMPROVER using ECMWF data, though future technical hurdles are far from ruled out.

## Next steps

I've rebased this long out of date branch on current master, and dealt with the resulting conflicts. The tests directly associated with basic thresholding all pass and have been slightly extended to cover vicinity processing. Enabling the plugin to once again handle single-valued thresholds has fixed almost all the other tests that were broken, this along with some small adaptations due to interface changes.

In summary, I believe this is now working, and as shown above in the suite test, will enable us to process ECMWF data without the need to balloon the memory requirements / significantly shrink the pool sizes.

I have opened the following IMPROVER PR: https://github.com/metoppv/improver/pull/1913

We will need to ensure this change is shared with collaborators (ideally there would be a BOM review) as it will change suite interfaces. I will raise a suite change to modify our threshold suite app next week.

**Comment by bayliffe** [comment]
@bjwheltor can you appraise these changes against your newly written guidance for metadata handling and make recommendations please. Specifically, the changes to functions in spatial.py which perform metadata manipulation regarding vicinity processing outside the context of a plugin, and do so piecemeal.

**Comment by bayliffe** [comment]
PR to factor out the percentile space checking. This is required as I need to modify the threshold plugin to be able to collapse a percentile coordinate, as has been done in the wind scaling implementation.

https://github.com/metoppv/improver/pull/1928

**Comment by bjwheltor** [comment]
> @bjwheltor can you appraise these changes against your newly written guidance for metadata handling and make recommendations please. Specifically, the changes to functions in spatial.py which perform metadata manipulation regarding vicinity processing outside the context of a plugin, and do so piecemeal.

Discussed this with @bayliffe and he main concern was that the principle is to keep metadata updating in the plug-ins, but here it is being pulled out  into a function. However, this just appears to be an implementation efficiency, as the plug-in is still actioning the metadata update, but just using the function, so I'm happy with this approach.

**Comment by gavinevans** [comment]
Just to validate the fact that the memory usage for the `threshold` step has reduced following https://github.com/metoppv/improver/pull/1913, I've run the following command, which shows a reduction in memory and an improvement to the timeliness of the CLI.

```
time bin/improver --memprofile=/scratch/gevans/improver_1787_efficient_threshold/profiling_master_ecinput threshold /scratch/gevans/improver_1787_efficient_threshold/20230926T1200Z-PT0012H00M-pressure_at_mean_sea_level.nc --output /scratch/gevans/improver_1787_efficient_threshold/output.nc --threshold-values 1000 --collapse-coord realization
time bin/improver --memprofile=/scratch/gevans/improver_1787_efficient_threshold/profiling_im1787_ecinput threshold /scratch/gevans/improver_1787_efficient_threshold/20230926T1200Z-PT0012H00M-pressure_at_mean_sea_level.nc --output /scratch/gevans/improver_1787_efficient_threshold/output.nc --threshold-values 1000 --collapse-coord realization
```

The results were:
|Experiment | Peak memory (MB) | Total run time (s) |
|---|---|---|
|PR 1913| 299 | 11.8 |
|master| 1021 | 12.3 |

Therefore the CLI appears to be reducing memory as intended.

**Comment by gavinevans** [comment]
I've now merged this PR: https://github.com/metoppv/improver/pull/1913, so this issue can be closed.
